### DISCUSSION TITLE
How to circumvent orm.Bundle contextual labels breaking serialization ?

### DISCUSSION THREAD

**Discussion by Neah-Ko** [sqlalchemy/sqlalchemy#11345] [Usage Questions] [answered]
Hello,

I am building some kind of REST-to-CRUD API framework built on top of SQLAlchemy and Marshmallow.
In order to allow for request optimization i am working on supporting a `?fields=f1,f2..fn` query string parameter from my incoming GET requests.

From the documentation, `Bundle` helps achieving that. I am demonstrating with `DictBundle` [class](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.Bundle) as it makes key-value pairs explicit.
At the moment the statement building looks like this (abstracting nested attribute inference and table elements accessing logic behind magic functions):

```python
def read(self, id, fields):
    attrs = get_non_nested_attributes(fields)
    nested = get_nested_attributes(fields)
    stmt = select(
        DictBundle(
           self. table.__name__, 
           *[getattr(self.table, a) for a in attrs],
           *[
                DictBundle(
                    nested_name, 
                    *[
                          getattr(get_target_table(self.table, nested_name), field)
                          for field in get_nested_fields(self.table, nested_name)
                     ]
                ),
                for nested_name in nested
            ]
        )
    )
    for nested_name in nested:
        stmt = stmt.join_from(self.table, get_target_table(self.table, nested_name)
    stmt = stmt.where(getattr(self.table, "id") == id)
    ....
```

Then this statement is processed through a `session.scalar(stmt)` and sent for serialization.
But then there are some missing fields.

E.g say I have a resource that belongs to a group in business logic, so the relationship is represented like this in a json dictionary.

```bash
curl -d '{"id": "1", "name": "test",  "somevalue": "1234",  "owner_group": {"name": "mygroup", "description": "lorem ipsum"}}' http://127.0.0.1:8000/myresources/
``` 

but then on retrieving data

```bash
curl http://127.0.0.1:8000/myresources/1?fields=name,owner_group
{"name": "test",   "owner_group": {"description": "lorem ipsum"}}
``` 

group `name` is missing.

When checking what data looks like before going into serialization:
``` 
"{'id': None, 'somevalue': None, 'name': 'test', 'owner_group': {'name_1': 'mygroup', 'description':'lorem ipsum', 'name_parent': None}}"
``` 

So here, we can see that data is well fetched from the DB but wrongly indexed for serialization.

I've tried to suppress contextual labeling by overloading `create_row_processor` for my `DictBundle` class but then that breaks data fetching.

It seems that if I want to keep using Bundles, I'm left with manually handling the dicts to apply table fields knowledge and the idea doesn't sounds terrific in lights of the performance gain I'm trying to offer.

  
Would you know of a proper way to generically fetch nested entities and partial results that can be serialized ? 

Best,

**Comment by zzzeek** [comment]
taht's a bug in select(), Query works.  issue upcoming

**Comment by zzzeek** [comment, marked as answer]
see #11347

**Reply by zzzeek** [reply]
there's not too much of a workaround besides using `session.query()` here at the moment, this fix will be out in the next release which should be within a few days

**Reply by Neah-Ko** [reply]
Awesome, I'll try `query` and stay on the lookout for the release. 

Thanks :star2:

**Reply by Neah-Ko** [reply]
Hi @zzzeek ,
Today, I tried with `query` and it indeed works.

However it is not available when using `AsyncSession`, I'll wait for the release.

Thanks again.
Best,
